core/app/models/spree/stock_location.rb

```diff
@@ -38,15 +38,13 @@ def state_text
     end
 
     # Wrapper for creating a new stock item respecting the backorderable config
-    def propagate_variant(variant)
-      stock_items.create!(variant: variant, backorderable: backorderable_default)
-    end
-
-    # Return either an existing stock item or create a new one. Useful in
-    # scenarios where the user might not know whether there is already a stock
-    # item for a given variant
-    def set_up_stock_item(variant)
-      stock_item(variant) || propagate_variant(variant)
+    def propagate_variant(variant_or_id)
+      vid = if variant_or_id.is_a?(Variant)
+        variant_or_id.id
+      else
+        variant_or_id
+      end
```

Comment:
I know that this would be a breaking change, but I wish that there was an elegant way to transition to using keyword arguments instead of multi-purpose arguments. I'd love this to be like:

```ruby
def propagate_variant(variant: nil, variant_id: nil)
  vid = variant_id || variant&.id or raise ArgumentError

  stock_items.create!(variant_id: vid, backorderable: backorderable_default)
end
```

One option could be to deprecate `propagate_variant` for a method called `propagate` (since I think it would still be clear due to the use of keyword args) and maintain `propagate_variant`'s API until it gets removed. Similarly we could create a `stock_item_for(variant: nil, variant_id: nil)` method.

I don't really feel strongly about this; I'm just not a fan of having these arguments that can be different things and inspecting them.

Reply:
I think this pattern of allowing both `id` or record is a bit unfortunate and results in the type checking we now have to have in multiple places in this class. We would have a much nicer API if we require a full `Variant` record for all public methods in this class that require it. If we are willing to deprecate anything it should be the ability to pass in `id` in `stock_item_or_create` so we can remove the type checking there, rather than propagating that throughout other existing methods. In my experience passing id's around usually results in more unnecessary queries because often the calling code already has access to a fully loaded record.

Reply:
Oh it's awful. I just kept it because it seemed like the least amount of change - absolutely happy to add `propagate` and deprecate the rest. I'd even go for a signature like `stock_item_for(variant: nil) signature; less stuff to worry about, and less mucking about with IDs that stop meaning anything up or down the callstack.